bc_obps/reporting/service/report_emission_allocation_service.py

```diff
@@ -0,0 +1,214 @@
+            if (
+                data["type"] == "other_excluded"
+            ):  # these categories cannot currently be allocated to products, so we skip them
+                continue
+            if (
+                data["type"] == "basic"
+            ):  # only these categories are reportable, so we sum them up to get the total reportable emissions
+                total_reportable_emissions += data["total"]
+            # Build product data
+            products = []
+            for rp in report_products:
+                product_emission = report_product_emission_allocations.filter(
+                    report_product_id=rp.id, emission_category__category_name=category
+                ).first()
+                product = ReportProductEmissionAllocationSchemaOut(
+                    report_product_id=rp.pk,
+                    product_name=rp.product.name,
+                    allocated_quantity=product_emission.allocated_quantity if product_emission else 0,
+                )
+                products.append(product)
+
+            # Add to the final response
+            emissions_total = ReportFacilityEmissionsSchemaOut(
+                emission_category=category,
+                products=products,
+                emission_total=data["total"],
+                category_type=data["type"],
+            )
+            report_product_emission_allocations_data.append(emissions_total)
+
+        return ReportProductEmissionAllocationsSchemaOut(
+            report_product_emission_allocations=report_product_emission_allocations_data,
+            facility_total_emissions=total_reportable_emissions,
+            report_product_emission_allocation_totals=ReportEmissionAllocationService.get_emission_totals_by_report_product(
+                facility_report_id
+            ),
+            allocation_methodology=allocation_methodology,
+            allocation_other_methodology_description=allocation_other_methodology_description,
+        )
+
+    @classmethod
+    @transaction.atomic()
+    def save_emission_allocation_data(
+        cls,
+        report_version_id: int,
+        facility_id: UUID,
+        data: ReportProductEmissionAllocationsSchemaIn,
+        user_guid: UUID,
+    ) -> None:
+
+        facility_report_id = FacilityReport.objects.get(report_version_id=report_version_id, facility_id=facility_id).pk
+
+        # Assumption: data is only entered when allocated_quantity is not 0
+        report_emission_allocations = data.report_product_emission_allocations
+        allocation_methodology = data.allocation_methodology
+        allocation_other_methodology_description = data.allocation_other_methodology_description
+
+        # Update or create the emission allocations from the data
+        for allocations in report_emission_allocations:
+            emission_category_id = EmissionCategory.objects.get(category_name=allocations.emission_category_name).pk
```

Comment:
We should probably carry the emission category Ids in the schema / form data as hidden fields & use them here rather than fetching by name. Fetching objects by name is fragile.